File: src/gazbean/GazBean.java
package edu.ucsb.alexandria.gazetteer;

import java.io.Serializable;
import java.sql.*;
import java.util.*;
import javax.sql.DataSource;

/**
 * ============================================================================
 * GazBean - Alexandria Digital Library (ADL) Gazetteer Component
 * ============================================================================
 * Java Bean component architecture designed for ADL / ADEPT georeferenced
 * digital libraries. Bridges web portals, GIS toolbars, and search middleware
 * to PostgreSQL / PostGIS gazetteer spatial databases and USGS GNIS services.
 *
 * Conforms to: ADL Gazetteer Content Standard (GCS v1.2)
 * Authors: Alexandria Digital Library Project, UC Santa Barbara
 * ============================================================================
 */
public class GazBean implements Serializable {
    private static final long serialVersionUID = 19990823L;

    // Database Connection Parameters
    private String jdbcUrl = "jdbc:postgresql://localhost:5432/adl_gazetteer";
    private String dbUser = "adl_user";
    private String dbPassword = "adl_password";
    private transient DataSource dataSource;

    // Query Criteria / State Variables
    private String placenameQuery;
    private String matchMode = "fuzzy"; // "exact", "fuzzy", "prefix", "fts"
    private double fuzzyThreshold = 0.35;
    private String fttCode; // ADL Feature Type Thesaurus code (e.g. "phys.peak")
    private String gnisFeatureClass;
    private String stateAlpha;
    private String countyName;

    // Spatial Footprint Filter (BBox & Radial)
    private Double minLon;
    private Double minLat;
    private Double maxLon;
    private Double maxLat;
    private Double centerLon;
    private Double centerLat;
    private Double radiusMeters;

    // Pagination & Sorting
    private int maxResults = 50;
    private int offset = 0;
    private String sortBy = "relevance"; // "relevance", "name", "distance", "elevation"

    // Default Constructor
    public GazBean() {}

    public GazBean(String jdbcUrl, String dbUser, String dbPassword) {
        this.jdbcUrl = jdbcUrl;
        this.dbUser = dbUser;
        this.dbPassword = dbPassword;
    }

    public GazBean(DataSource dataSource) {
        this.dataSource = dataSource;
    }

    // ========================================================================
    // GETTERS & SETTERS (JavaBean Spec)
    // ========================================================================
    public String getPlacenameQuery() { return placenameQuery; }
    public void setPlacenameQuery(String placenameQuery) { this.placenameQuery = placenameQuery; }

    public String getMatchMode() { return matchMode; }
    public void setMatchMode(String matchMode) { this.matchMode = matchMode; }

    public double getFuzzyThreshold() { return fuzzyThreshold; }
    public void setFuzzyThreshold(double fuzzyThreshold) { this.fuzzyThreshold = fuzzyThreshold; }

    public String getFttCode() { return fttCode; }
    public void setFttCode(String fttCode) { this.fttCode = fttCode; }

    public String getGnisFeatureClass() { return gnisFeatureClass; }
    public void setGnisFeatureClass(String gnisFeatureClass) { this.gnisFeatureClass = gnisFeatureClass; }

    public String getStateAlpha() { return stateAlpha; }
    public void setStateAlpha(String stateAlpha) { this.stateAlpha = stateAlpha; }

    public String getCountyName() { return countyName; }
    public void setCountyName(String countyName) { this.countyName = countyName; }

    public void setBoundingBox(double minLon, double minLat, double maxLon, double maxLat) {
        this.minLon = minLon;
        this.minLat = minLat;
        this.maxLon = maxLon;
        this.maxLat = maxLat;
    }

    public void setRadialFilter(double centerLon, double centerLat, double radiusMeters) {
        this.centerLon = centerLon;
        this.centerLat = centerLat;
        this.radiusMeters = radiusMeters;
    }

    public int getMaxResults() { return maxResults; }
    public void setMaxResults(int maxResults) { this.maxResults = maxResults; }

    public int getOffset() { return offset; }
    public void setOffset(int offset) { this.offset = offset; }

    public String getSortBy() { return sortBy; }
    public void setSortBy(String sortBy) { this.sortBy = sortBy; }

    // ========================================================================
    // DATABASE CONNECTION MANAGEMENT
    // ========================================================================
    protected Connection getConnection() throws SQLException {
        if (this.dataSource != null) {
            return this.dataSource.getConnection();
        }
        return DriverManager.getConnection(this.jdbcUrl, this.dbUser, this.dbPassword);
    }

    // ========================================================================
    // QUERY EXECUTION & POSTGIS SQL GENERATION
    // ========================================================================
    
    /**
     * Executes the configured spatial gazetteer query against PostgreSQL/PostGIS.
     * Returns a list of matching GazetteerFeature objects conforming to ADL GCS.
     */
    public List<GazetteerFeature> executeQuery() throws SQLException {
        List<GazetteerFeature> results = new ArrayList<>();
        StringBuilder sql = new StringBuilder();
        List<Object> params = new ArrayList<>();

        sql.append("SELECT f.adl_id, f.gnis_id, f.primary_name, f.ftt_code, f.gnis_feature_class, ")
           .append("f.state_alpha, f.county_name, f.ele_in_m, f.ele_in_ft, f.usgs_quad_name, ")
           .append("ST_Y(f.geom_point) AS lat, ST_X(f.geom_point) AS lon, ")
           .append("ST_AsGeoJSON(f.geom_point) AS geojson_point, ")
           .append("ST_AsGeoJSON(f.geom_bbox) AS geojson_bbox, ")
           .append("f.metadata_json ");

        if (centerLon != null && centerLat != null) {
            sql.append(", ST_Distance(f.geog_point, ST_SetSRID(ST_MakePoint(?, ?), 4326)::geography) AS distance_meters ");
            params.add(centerLon);
            params.add(centerLat);
        }

        sql.append("FROM adl_gazetteer_features f ");
        sql.append("WHERE 1=1 ");

        // Spatial Bounding Box Filter using PostGIS R-Tree GiST operator (&&) + ST_Intersects
        if (minLon != null && minLat != null && maxLon != null && maxLat != null) {
            sql.append("AND f.geom_point && ST_MakeEnvelope(?, ?, ?, ?, 4326) ")
               .append("AND ST_Intersects(f.geom_point, ST_MakeEnvelope(?, ?, ?, ?, 4326)) ");
            params.add(minLon); params.add(minLat); params.add(maxLon); params.add(maxLat);
            params.add(minLon); params.add(minLat); params.add(maxLon); params.add(maxLat);
        }

        // Radial Geodesic Filter using PostGIS ST_DWithin on geography
        if (centerLon != null && centerLat != null && radiusMeters != null) {
            sql.append("AND ST_DWithin(f.geog_point, ST_SetSRID(ST_MakePoint(?, ?), 4326)::geography, ?) ");
            params.add(centerLon);
            params.add(centerLat);
            params.add(radiusMeters);
        }

        // Placename Filter (Fuzzy Trigram, Exact, Prefix, or FTS)
        if (placenameQuery != null && !placenameQuery.trim().isEmpty()) {
            String q = placenameQuery.trim().toLowerCase();
            switch (matchMode.toLowerCase()) {
                case "exact":
                    sql.append("AND LOWER(f.primary_name) = ? ");
                    params.add(q);
                    break;
                case "prefix":
                    sql.append("AND LOWER(f.primary_name) LIKE ? ");
                    params.add(q + "%");
                    break;
                case "fts":
                    sql.append("AND to_tsvector('english', f.primary_name) @@ plainto_tsquery('english', ?) ");
                    params.add(q);
                    break;
                case "fuzzy":
                default:
                    sql.append("AND (f.normalized_name % ? OR similarity(f.normalized_name, ?) >= ? OR f.primary_name ILIKE ?) ");
                    params.add(q);
                    params.add(q);
                    params.add(fuzzyThreshold);
                    params.add("%" + q + "%");
                    break;
            }
        }

        // Feature Classification Filter (ADL FTT hierarchy)
        if (fttCode != null && !fttCode.isEmpty()) {
            sql.append("AND (f.ftt_code = ? OR f.ftt_code LIKE ?) ");
            params.add(fttCode);
            params.add(fttCode + ".%");
        }

        if (gnisFeatureClass != null && !gnisFeatureClass.isEmpty()) {
            sql.append("AND f.gnis_feature_class = ? ");
            params.add(gnisFeatureClass);
        }

        // State and County Filter
        if (stateAlpha != null && !stateAlpha.isEmpty()) {
            sql.append("AND f.state_alpha = ? ");
            params.add(stateAlpha.toUpperCase());
        }
        if (countyName != null && !countyName.isEmpty()) {
            sql.append("AND LOWER(f.county_name) = ? ");
            params.add(countyName.toLowerCase());
        }

        // Ordering & Pagination
        if ("distance".equalsIgnoreCase(sortBy) && centerLon != null && centerLat != null) {
            sql.append("ORDER BY distance_meters ASC ");
        } else if ("elevation".equalsIgnoreCase(sortBy)) {
            sql.append("ORDER BY f.ele_in_m DESC NULLS LAST ");
        } else {
            sql.append("ORDER BY f.primary_name ASC ");
        }

        sql.append("LIMIT ? OFFSET ?");
        params.add(maxResults);
        params.add(offset);

        try (Connection conn = getConnection();
             PreparedStatement ps = conn.prepareStatement(sql.toString())) {
            for (int i = 0; i < params.size(); i++) {
                ps.setObject(i + 1, params.get(i));
            }

            try (ResultSet rs = ps.executeQuery()) {
                while (rs.next()) {
                    GazetteerFeature feat = new GazetteerFeature();
                    feat.setAdlId(rs.getString("adl_id"));
                    feat.setGnisId(rs.getLong("gnis_id"));
                    feat.setPrimaryName(rs.getString("primary_name"));
                    feat.setFttCode(rs.getString("ftt_code"));
                    feat.setGnisFeatureClass(rs.getString("gnis_feature_class"));
                    feat.setStateAlpha(rs.getString("state_alpha"));
                    feat.setCountyName(rs.getString("county_name"));
                    feat.setElevationMeters(rs.getDouble("ele_in_m"));
                    feat.setElevationFeet(rs.getDouble("ele_in_ft"));
                    feat.setUsgsQuadName(rs.getString("usgs_quad_name"));
                    feat.setLatitude(rs.getDouble("lat"));
                    feat.setLongitude(rs.getDouble("lon"));
                    feat.setGeoJsonPoint(rs.getString("geojson_point"));
                    feat.setGeoJsonBBox(rs.getString("geojson_bbox"));
                    feat.setMetadataJson(rs.getString("metadata_json"));
                    results.add(feat);
                }
            }
        }

        // Populate Variant Names for each feature
        if (!results.isEmpty()) {
            attachVariantNames(results);
        }

        return results;
    }

    private void attachVariantNames(List<GazetteerFeature> features) throws SQLException {
        if (features.isEmpty()) return;
        Map<String, GazetteerFeature> featureMap = new HashMap<>();
        List<String> ids = new ArrayList<>();
        for (GazetteerFeature f : features) {
            featureMap.put(f.getAdlId(), f);
            ids.add("?");
        }

        String sql = "SELECT adl_id, variant_name, name_type, language_iso FROM adl_variant_names WHERE adl_id IN (" 
                   + String.join(",", ids) + ")";

        try (Connection conn = getConnection();
             PreparedStatement ps = conn.prepareStatement(sql)) {
            for (int i = 0; i < ids.size(); i++) {
                ps.setString(i + 1, features.get(i).getAdlId());
            }
            try (ResultSet rs = ps.executeQuery()) {
                while (rs.next()) {
                    String adlId = rs.getString("adl_id");
                    GazetteerFeature feat = featureMap.get(adlId);
                    if (feat != null) {
                        feat.addVariantName(rs.getString("variant_name"));
                    }
                }
            }
        }
    }

    // ========================================================================
    // ADL GAZETTEER DATA TRANSFER OBJECT (GCS DTO)
    // ========================================================================
    public static class GazetteerFeature implements Serializable {
        private static final long serialVersionUID = 1L;
        private String adlId;
        private Long gnisId;
        private String primaryName;
        private List<String> variantNames = new ArrayList<>();
        private String fttCode;
        private String gnisFeatureClass;
        private String stateAlpha;
        private String countyName;
        private Double elevationMeters;
        private Double elevationFeet;
        private String usgsQuadName;
        private double latitude;
        private double longitude;
        private String geoJsonPoint;
        private String geoJsonBBox;
        private String metadataJson;

        public String getAdlId() { return adlId; }
        public void setAdlId(String adlId) { this.adlId = adlId; }

        public Long getGnisId() { return gnisId; }
        public void setGnisId(Long gnisId) { this.gnisId = gnisId; }

        public String getPrimaryName() { return primaryName; }
        public void setPrimaryName(String primaryName) { this.primaryName = primaryName; }

        public List<String> getVariantNames() { return variantNames; }
        public void addVariantName(String variant) { this.variantNames.add(variant); }

        public String getFttCode() { return fttCode; }
        public void setFttCode(String fttCode) { this.fttCode = fttCode; }

        public String getGnisFeatureClass() { return gnisFeatureClass; }
        public void setGnisFeatureClass(String gnisFeatureClass) { this.gnisFeatureClass = gnisFeatureClass; }

        public String getStateAlpha() { return stateAlpha; }
        public void setStateAlpha(String stateAlpha) { this.stateAlpha = stateAlpha; }

        public String getCountyName() { return countyName; }
        public void setCountyName(String countyName) { this.countyName = countyName; }

        public Double getElevationMeters() { return elevationMeters; }
        public void setElevationMeters(Double elevationMeters) { this.elevationMeters = elevationMeters; }

        public Double getElevationFeet() { return elevationFeet; }
        public void setElevationFeet(Double elevationFeet) { this.elevationFeet = elevationFeet; }

        public String getUsgsQuadName() { return usgsQuadName; }
        public void setUsgsQuadName(String usgsQuadName) { this.usgsQuadName = usgsQuadName; }

        public double getLatitude() { return latitude; }
        public void setLatitude(double latitude) { this.latitude = latitude; }

        public double getLongitude() { return longitude; }
        public void setLongitude(double longitude) { this.longitude = longitude; }

        public String getGeoJsonPoint() { return geoJsonPoint; }
        public void setGeoJsonPoint(String geoJsonPoint) { this.geoJsonPoint = geoJsonPoint; }

        public String getGeoJsonBBox() { return geoJsonBBox; }
        public void setGeoJsonBBox(String geoJsonBBox) { this.geoJsonBBox = geoJsonBBox; }

        public String getMetadataJson() { return metadataJson; }
        public void setMetadataJson(String metadataJson) { this.metadataJson = metadataJson; }

        /**
         * Serializes the feature into ADL Gazetteer Content Standard XML.
         */
        public String toAdlXml() {
            StringBuilder xml = new StringBuilder();
            xml.append("<gazetteer-entry identifier=\"").append(adlId).append("\" gnis-id=\"").append(gnisId).append("\">\n")
               .append("  <names>\n")
               .append("    <primary-name>").append(escapeXml(primaryName)).append("</primary-name>\n");
            for (String v : variantNames) {
                xml.append("    <variant-name>").append(escapeXml(v)).append("</variant-name>\n");
            }
            xml.append("  </names>\n")
               .append("  <classification ftt-code=\"").append(fttCode).append("\">")
               .append(escapeXml(gnisFeatureClass)).append("</classification>\n")
               .append("  <spatial-footprint srid=\"EPSG:4326\">\n")
               .append("    <point latitude=\"").append(latitude).append("\" longitude=\"").append(longitude).append("\"/>\n");
            if (elevationMeters != null) {
                xml.append("    <elevation units=\"meters\">").append(elevationMeters).append("</elevation>\n");
            }
            xml.append("  </spatial-footprint>\n")
               .append("  <administrative-areas>\n")
               .append("    <state>").append(stateAlpha).append("</state>\n")
               .append("    <county>").append(escapeXml(countyName)).append("</county>\n")
               .append("    <usgs-quad>").append(escapeXml(usgsQuadName)).append("</usgs-quad>\n")
               .append("  </administrative-areas>\n")
               .append("</gazetteer-entry>");
            return xml.toString();
        }

        private String escapeXml(String text) {
            if (text == null) return "";
            return text.replace("&", "&amp;").replace("<", "&lt;").replace(">", "&gt;").replace("\"", "&quot;");
        }
    }
}

File: src/gazbean/gazbean.ts
/**
 * ============================================================================
 * GazBean TypeScript / ES6 - Alexandria Digital Library Gazetteer Module
 * ============================================================================
 * Universal TypeScript/JavaScript implementation of GazBean.
 * Powers the interactive HTML5 spatial library explorer, generates PostGIS SQL,
 * executes spatial queries, and transforms USGS GNIS data to ADL GCS v1.2 records.
 * ============================================================================
 */

export interface BoundingBox {
  minLon: number;
  minLat: number;
  maxLon: number;
  maxLat: number;
}

export interface RadialPoint {
  lon: number;
  lat: number;
  radiusMeters: number;
}

export interface GazetteerFeature {
  adlId: string;
  gnisId: number;
  primaryName: string;
  normalizedName: string;
  fttCode: string;
  gnisFeatureClass: string;
  stateAlpha: string;
  countyName: string;
  latitude: number;
  longitude: number;
  elevationMeters?: number;
  elevationFeet?: number;
  usgsQuadName?: string;
  bbox?: [number, number, number, number];
  variantNames: string[];
  metadata: Record<string, any>;
  distanceMeters?: number;
}

export interface QueryOptions {
  name?: string;
  matchMode?: "exact" | "prefix" | "fuzzy" | "fts";
  fttCode?: string;
  gnisClass?: string;
  state?: string;
  county?: string;
  bbox?: BoundingBox;
  radial?: RadialPoint;
  minEleMeters?: number;
  maxEleMeters?: number;
  limit?: number;
  offset?: number;
  sortBy?: "name" | "distance" | "elevation";
}

/**
 * Calculates geodesic distance between two coordinate pairs in meters (Haversine formula).
 */
export function haversineDistance(lat1: number, lon1: number, lat2: number, lon2: number): number {
  const R = 6371008.8; // meters
  const dLat = ((lat2 - lat1) * Math.PI) / 180;
  const dLon = ((lon2 - lon1) * Math.PI) / 180;
  const a =
    Math.sin(dLat / 2) * Math.sin(dLat / 2) +
    Math.cos((lat1 * Math.PI) / 180) *
      Math.cos((lat2 * Math.PI) / 180) *
      Math.sin(dLon / 2) *
      Math.sin(dLon / 2);
  const c = 2 * Math.atan2(Math.sqrt(a), Math.sqrt(1 - a));
  return R * c;
}

/**
 * ADL Feature Type Thesaurus (FTT) Hierarchical Facets
 */
export const ADL_FTT_TAXONOMY: Record<string, { name: string; facet: string; gnis: string[] }> = {
  "phys": { name: "Physiographic Features", facet: "Physiographic", gnis: ["Summit", "Valley", "Ridge", "Cliff", "Range", "Island"] },
  "phys.peak": { name: "Peaks & Summits", facet: "Physiographic", gnis: ["Summit", "Pillar"] },
  "phys.valley": { name: "Valleys & Canyons", facet: "Physiographic", gnis: ["Valley", "Canyon", "Gulch", "Arroyo"] },
  "phys.range": { name: "Mountain Ranges", facet: "Physiographic", gnis: ["Range", "Ridge"] },
  "phys.island": { name: "Islands & Reefs", facet: "Physiographic", gnis: ["Island", "Bar", "Reef"] },
  "hydro": { name: "Hydrographic Structures", facet: "Hydrographic", gnis: ["Stream", "Lake", "Spring", "Bay", "Canal", "Reservoir"] },
  "hydro.stream": { name: "Streams & Rivers", facet: "Hydrographic", gnis: ["Stream", "River", "Creek", "Brook"] },
  "hydro.lake": { name: "Lakes & Ponds", facet: "Hydrographic", gnis: ["Lake", "Pond", "Reservoir"] },
  "hydro.marine": { name: "Marine Coastal Features", facet: "Hydrographic", gnis: ["Bay", "Channel", "Sound", "Harbor"] },
  "hydro.spring": { name: "Springs & Wells", facet: "Hydrographic", gnis: ["Spring", "Well", "Geyser"] },
  "pop": { name: "Populated Places", facet: "Populated Places", gnis: ["Populated Place", "Locale"] },
  "pop.city": { name: "Incorporated Cities", facet: "Populated Places", gnis: ["Populated Place"] },
  "pop.town": { name: "Unincorporated Towns", facet: "Populated Places", gnis: ["Populated Place", "Locale"] },
  "admin": { name: "Administrative Areas", facet: "Administrative", gnis: ["Civil", "Reserve", "Park", "Forest", "Military"] },
  "admin.park": { name: "Parks & Reserves", facet: "Administrative", gnis: ["Park", "Reserve", "Forest"] },
  "admin.civil": { name: "Civil Divisions", facet: "Administrative", gnis: ["Civil"] },
  "admin.military": { name: "Military Installations", facet: "Administrative", gnis: ["Military"] },
  "manmade": { name: "Manmade Structures", facet: "Manmade", gnis: ["Building", "Tower", "Bridge", "Dam", "Airport", "School"] },
  "manmade.transport": { name: "Transportation Facilities", facet: "Manmade", gnis: ["Airport", "Bridge", "Canal", "Crossing"] }
};

export class GazBeanEngine {
  private records: GazetteerFeature[] = [];

  constructor(initialRecords: GazetteerFeature[] = []) {
    this.records = [...initialRecords];
  }

  public addFeature(feature: GazetteerFeature): void {
    this.records.push(feature);
  }

  public setFeatures(features: GazetteerFeature[]): void {
    this.records = [...features];
  }

  public getFeatures(): GazetteerFeature[] {
    return this.records;
  }

  /**
   * Generates production PostGIS SQL query reflecting current parameters.
   */
  public generatePostGisSql(opts: QueryOptions): { sql: string; explanation: string } {
    const lines: string[] = [
      "-- GazBean PostGIS Spatial Query Planner",
      "SELECT",
      "    f.adl_id,",
      "    f.gnis_id,",
      "    f.primary_name,",
      "    f.ftt_code,",
      "    f.gnis_feature_class,",
      "    f.state_alpha,",
      "    f.county_name,",
      "    f.ele_in_m,",
      "    f.usgs_quad_name,",
      "    ST_Y(f.geom_point) AS lat,",
      "    ST_X(f.geom_point) AS lon,",
      "    ST_AsGeoJSON(f.geom_point) AS geojson"
    ];

    if (opts.radial) {
      lines.push(`    , ST_Distance(f.geog_point, ST_SetSRID(ST_MakePoint(${opts.radial.lon}, ${opts.radial.lat}), 4326)::geography) AS distance_meters`);
    }

    lines.push("FROM adl_gazetteer_features f");
    lines.push("WHERE 1=1");

    const explanations: string[] = [];

    if (opts.bbox) {
      const b = opts.bbox;
      lines.push(`  -- Spatial Bounding Box Filter (Indexed via GiST R-Tree)`);
      lines.push(`  AND f.geom_point && ST_MakeEnvelope(${b.minLon.toFixed(4)}, ${b.minLat.toFixed(4)}, ${b.maxLon.toFixed(4)}, ${b.maxLat.toFixed(4)}, 4326)`);
      lines.push(`  AND ST_Intersects(f.geom_point, ST_MakeEnvelope(${b.minLon.toFixed(4)}, ${b.minLat.toFixed(4)}, ${b.maxLon.toFixed(4)}, ${b.maxLat.toFixed(4)}, 4326))`);
      explanations.push(`Spatial bounding box filter: [${b.minLon.toFixed(3)}, ${b.minLat.toFixed(3)} to ${b.maxLon.toFixed(3)}, ${b.maxLat.toFixed(3)}]`);
    }

    if (opts.radial) {
      const r = opts.radial;
      lines.push(`  -- Radial Geodesic Filter on WGS84 Spheroid`);
      lines.push(`  AND ST_DWithin(f.geog_point, ST_SetSRID(ST_MakePoint(${r.lon.toFixed(4)}, ${r.lat.toFixed(4)}), 4326)::geography, ${r.radiusMeters})`);
      explanations.push(`Radial filter: ${r.radiusMeters}m buffer around (${r.lat.toFixed(4)}, ${r.lon.toFixed(4)})`);
    }

    if (opts.name && opts.name.trim()) {
      const q = opts.name.trim().replace(/'/g, "''");
      const mode = opts.matchMode || "fuzzy";
      if (mode === "exact") {
        lines.push(`  AND LOWER(f.primary_name) = '${q.toLowerCase()}'`);
        explanations.push(`Exact placename match for '${q}'`);
      } else if (mode === "prefix") {
        lines.push(`  AND LOWER(f.primary_name) LIKE '${q.toLowerCase()}%'`);
        explanations.push(`Prefix placename match for '${q}%'`);
      } else if (mode === "fts") {
        lines.push(`  AND to_tsvector('english', f.primary_name) @@ plainto_tsquery('english', '${q}')`);
        explanations.push(`PostgreSQL GIN Full-Text Search for '${q}'`);
      } else {
        lines.push(`  -- Trigram similarity matching using pg_trgm GIN/GiST index`);
        lines.push(`  AND (f.normalized_name % '${q.toLowerCase()}' OR similarity(f.normalized_name, '${q.toLowerCase()}') >= 0.35 OR f.primary_name ILIKE '%${q}%')`);
        explanations.push(`Fuzzy trigram similarity threshold >= 0.35 for '${q}'`);
      }
    }

    if (opts.fttCode) {
      lines.push(`  -- ADL Feature Type Thesaurus hierarchy containment`);
      lines.push(`  AND (f.ftt_code = '${opts.fttCode}' OR f.ftt_code LIKE '${opts.fttCode}.%')`);
      explanations.push(`ADL FTT category filter: '${opts.fttCode}'`);
    }

    if (opts.gnisClass) {
      lines.push(`  AND f.gnis_feature_class = '${opts.gnisClass}'`);
      explanations.push(`USGS GNIS class filter: '${opts.gnisClass}'`);
    }

    if (opts.state) {
      lines.push(`  AND f.state_alpha = '${opts.state.toUpperCase()}'`);
      explanations.push(`State filter: '${opts.state.toUpperCase()}'`);
    }

    if (opts.county) {
      lines.push(`  AND LOWER(f.county_name) = '${opts.county.toLowerCase()}'`);
      explanations.push(`County filter: '${opts.county}'`);
    }

    if (opts.minEleMeters !== undefined) {
      lines.push(`  AND f.ele_in_m >= ${opts.minEleMeters}`);
    }
    if (opts.maxEleMeters !== undefined) {
      lines.push(`  AND f.ele_in_m <= ${opts.maxEleMeters}`);
    }

    if (opts.radial || opts.sortBy === "distance") {
      lines.push("ORDER BY distance_meters ASC");
    } else if (opts.sortBy === "elevation") {
      lines.push("ORDER BY f.ele_in_m DESC NULLS LAST");
    } else {
      lines.push("ORDER BY f.primary_name ASC");
    }

    lines.push(`LIMIT ${opts.limit || 50} OFFSET ${opts.offset || 0};`);

    return {
      sql: lines.join("\n"),
      explanation: explanations.join(" | ") || "Full gazetteer table scan / index scan"
    };
  }

  /**
   * Executes in-browser spatial gazetteer query.
   */
  public query(opts: QueryOptions): GazetteerFeature[] {
    const results: GazetteerFeature[] = [];
    const limit = opts.limit || 50;
    const offset = opts.offset || 0;

    for (const feat of this.records) {
      // Spatial BBox
      if (opts.bbox) {
        const b = opts.bbox;
        if (
          feat.longitude < b.minLon ||
          feat.longitude > b.maxLon ||
          feat.latitude < b.minLat ||
          feat.latitude > b.maxLat
        ) {
          continue;
        }
      }

      // Radial Distance
      if (opts.radial) {
        const r = opts.radial;
        const dist = haversineDistance(r.lat, r.lon, feat.latitude, feat.longitude);
        if (dist > r.radiusMeters) continue;
        feat.distanceMeters = dist;
      } else {
        feat.distanceMeters = undefined;
      }

      // Placename search
      if (opts.name && opts.name.trim()) {
        const q = opts.name.trim().toLowerCase();
        let match = false;
        const norm = feat.normalizedName || feat.primaryName.toLowerCase();
        const mode = opts.matchMode || "fuzzy";

        if (mode === "exact") {
          if (norm === q || feat.variantNames.some((v) => v.toLowerCase() === q)) match = true;
        } else if (mode === "prefix") {
          if (norm.startsWith(q) || feat.variantNames.some((v) => v.toLowerCase().startsWith(q))) match = true;
        } else {
          // fuzzy / substring / token match
          if (norm.includes(q) || feat.variantNames.some((v) => v.toLowerCase().includes(q))) {
            match = true;
          } else {
            const tokens = norm.split(/\s+/);
            if (tokens.some((t) => t.startsWith(q))) match = true;
          }
        }

        if (!match) continue;
      }

      // FTT Code
      if (opts.fttCode) {
        if (feat.fttCode !== opts.fttCode && !feat.fttCode.startsWith(opts.fttCode + ".")) {
          continue;
        }
      }

      // GNIS Class
      if (opts.gnisClass && feat.gnisFeatureClass.toLowerCase() !== opts.gnisClass.toLowerCase()) {
        continue;
      }

      // State
      if (opts.state && feat.stateAlpha.toUpperCase() !== opts.state.toUpperCase()) {
        continue;
      }

      // County
      if (opts.county && (!feat.countyName || !feat.countyName.toLowerCase().includes(opts.county.toLowerCase()))) {
        continue;
      }

      // Elevation range
      if (opts.minEleMeters !== undefined && (feat.elevationMeters === undefined || feat.elevationMeters < opts.minEleMeters)) {
        continue;
      }
      if (opts.maxEleMeters !== undefined && (feat.elevationMeters === undefined || feat.elevationMeters > opts.maxEleMeters)) {
        continue;
      }

      results.push(feat);
    }

    // Sort
    if (opts.radial || opts.sortBy === "distance") {
      results.sort((a, b) => (a.distanceMeters || 0) - (b.distanceMeters || 0));
    } else if (opts.sortBy === "elevation") {
      results.sort((a, b) => (b.elevationMeters || 0) - (a.elevationMeters || 0));
    } else {
      results.sort((a, b) => a.primaryName.localeCompare(b.primaryName));
    }

    return results.slice(offset, offset + limit);
  }

  /**
   * Generates GeoJSON FeatureCollection.
   */
  public toGeoJson(features: GazetteerFeature[]): any {
    return {
      type: "FeatureCollection",
      metadata: {
        count: features.length,
        standard: "ADL Gazetteer Content Standard v1.2",
        generator: "GazBean Engine HTML5"
      },
      features: features.map((f) => ({
        type: "Feature",
        id: f.adlId,
        geometry: {
          type: "Point",
          coordinates: [f.longitude, f.latitude]
        },
        properties: {
          adl_id: f.adlId,
          gnis_id: f.gnisId,
          primary_name: f.primaryName,
          variant_names: f.variantNames,
          ftt_code: f.fttCode,
          gnis_class: f.gnisFeatureClass,
          state: f.stateAlpha,
          county: f.countyName,
          elevation_m: f.elevationMeters,
          elevation_ft: f.elevationFeet,
          usgs_quad: f.usgsQuadName,
          distance_meters: f.distanceMeters ? Math.round(f.distanceMeters) : undefined,
          metadata: f.metadata
        }
      }))
    };
  }

  /**
   * Generates ADL GCS v1.2 XML document.
   */
  public toAdlXml(features: GazetteerFeature[]): string {
    const entries = features
      .map(
        (f) => `  <gazetteer-entry identifier="${f.adlId}" gnis-id="${f.gnisId}">
    <names>
      <primary-name>${escapeXml(f.primaryName)}</primary-name>
      ${f.variantNames.map((v) => `<variant-name>${escapeXml(v)}</variant-name>`).join("\n      ")}
    </names>
    <classification ftt-code="${f.fttCode}">${escapeXml(f.gnisFeatureClass)}</classification>
    <spatial-footprint srid="EPSG:4326">
      <point latitude="${f.latitude}" longitude="${f.longitude}"/>
      ${f.elevationMeters !== undefined ? `<elevation units="meters">${f.elevationMeters}</elevation>` : ""}
    </spatial-footprint>
    <administrative-areas>
      <country>USA</country>
      <state>${f.stateAlpha}</state>
      <county>${escapeXml(f.countyName || "")}</county>
      <usgs-quadrangle>${escapeXml(f.usgsQuadName || "")}</usgs-quadrangle>
    </administrative-areas>
  </gazetteer-entry>`
      )
      .join("\n");

    return `<?xml version="1.0" encoding="UTF-8"?>\n<adl-gazetteer-response count="${features.length}" standard="ADL-GCS-1.2">\n${entries}\n</adl-gazetteer-response>`;
  }
}

function escapeXml(str: string): string {
  if (!str) return "";
  return str.replace(/&/g, "&amp;").replace(/</g, "&lt;").replace(/>/g, "&gt;").replace(/"/g, "&quot;");
}
